src/install/lockfile.zig

```diff
@@ -3853,3 +3699,1012 @@ pub fn resolve(this: *Lockfile, package_name: []const u8, version: Dependency.Ve
+
+                // Write the path into the same buffer we will read the file into
+                bun.copy(u8, body_pool.data.list.items, "node_modules" ++ std.fs.path.sep_str);
+                bun.copy(u8, body_pool.data.list.items[node_modules_offset..], name);
+                bun.copy(u8, body_pool.data.list.items[name_offset..], std.fs.path.sep_str ++ "package.json");
+                body_pool.data.list.items[path_len] = 0;
+
+                const package_json_file = bun.openFileZ(body_pool.data.list.items[0..path_len :0], .{}) catch |err| {
+                    Output.prettyErrorln("<r><red>error<r>: <b><red>{s}<r> opening <b>{s}<r> folder", .{ @errorName(err), body_pool.data.list.items });
+                    Global.crash();
+                };
+
+                const file_read_error = body_pool.data.readFile(package_json_file, path_len);
+                package_json_file.close();
+                try file_read_error;
+
+                const file_size = body_pool.data.list.items.len;
+                var package_json_path = body_pool.data.list.items.ptr[file_size .. file_size + path_len];
+
+                // Write the path into the same buffer we already read the file into.
+                // Kinda dumb but we can remove this at some point
+                bun.copy(u8, package_json_path, "node_modules" ++ std.fs.path.sep_str);
+                bun.copy(u8, package_json_path[node_modules_offset..], name);
+                bun.copy(u8, package_json_path[name_offset..], std.fs.path.sep_str ++ "package.json");
+
+                const source = logger.Source.initPathString(package_json_path, body_pool.data.list.items);
+
+                var log = logger.Log.init(allocator);
+                defer log.deinit();
+
+                const BufferWriter = std.io.FixedBufferStream([]u8);
+                var version_writer = BufferWriter{ .buffer = &version_buffer, .pos = 0 };
+
+                initializeStore();
+                var package_json_checker = try json_parser.PackageJSONVersionChecker.init(allocator, &source, &log);
+                package_json_checker.has_found_name = true; // ignore the name field
+                _ = try package_json_checker.parseExpr(std.io.null_writer, version_writer.writer());
+                // pass in our writer to write the version
+                if (log.errors > 0) {
+                    Output.prettyErrorln("<r><red>error<r>: <b><red>Invalid package.json<r>: <b>{s}<r> folder", .{body_pool.data.list.items});
+                    Global.crash();
+                }
+
+                // This parser is kinda overkill, but w/e
+                const ver = Semver.Version.parse(.{
+                    .buf = &version_buffer,
+                    .slice = version_writer.getWritten(),
+                }, allocator);
+
+                break :version_formatter if (ver.stopped_at < version_writer.pos or !ver.valid or ver.wildcard != .none)
```

Comment:
The `stopped_at` field is a bit wild. It can be one or two beyond the position of the last character.